internal/controller/providers/psmdb/applier.go

```diff
@@ -343,6 +327,117 @@ func (p *applier) Monitoring() error {
 	return nil
 }
 
+func (p *applier) PodSchedulingPolicy(systemNamespace string) error {
+	// NOTE: this method shall be called after Engine() and Proxy() methods
+	// because it extends the engine and proxy specs with the affinity rules.
+	//
+	// The following cases are possible:
+	// 1. The user did not specify a .spec.podSchedulingPolicyName -> do nothing.
+	// 2. The user specified a .spec.podSchedulingPolicyName, but it does not exist -> return error.
+	// 3. The user specified a .spec.podSchedulingPolicyName, and it exists, but it is not applicable to the upstream cluster -> return error.
+	// 4. The user specified a .spec.podSchedulingPolicyName, and it exists, but affinity configuration is absent there -> do nothing.
+	// 5. The user specified a .spec.podSchedulingPolicyName, and it exists, and it is applicable to the upstream cluster ->
+	// copy the affinity rules to the upstream cluster spec from policy.
+
+	psmdb := p.PerconaServerMongoDB
+	// --------------------------------- //
+	// Special workaround, need to reset all affinity params in p.PerconaServerMongoDB before moving further.
+	// TODO: Remove it once https://perconadev.atlassian.net/browse/EVEREST-2023 is addressed
+	for i := range len(psmdb.Spec.Replsets) {
+		psmdb.Spec.Replsets[i].MultiAZ.Affinity = nil
+	}
+
+	if psmdb.Spec.Sharding.Enabled {
+		// Config Server
+		psmdb.Spec.Sharding.ConfigsvrReplSet.MultiAZ.Affinity = nil
+		// Proxy
+		psmdb.Spec.Sharding.Mongos.MultiAZ.Affinity = nil
+	}
+	// --------------------------------- //
+	pspName := p.DB.Spec.PodSchedulingPolicyName
+
+	if pspName == "" {
+		// Covers case 1.
+
+		// We preserve the settings for existing DBs, otherwise restarts are seen when upgrading Everest.
+		// TODO: Remove this once we figure out how to apply such spec changes without automatic restarts.
+		// See: https://perconadev.atlassian.net/browse/EVEREST-1413
+
+		// if p.DB.Status.Status == everestv1alpha1.AppStateReady {
+		// 	// ReplicaSets
+		//
+		// 	for i := range len(psmdb.Spec.Replsets) {
+		// 		// all replsets have the same affinity config.
+		// 		psmdb.Spec.Replsets[i].MultiAZ.Affinity = p.currentPSMDBSpec.Replsets[0].MultiAZ.Affinity
+		// 	}
+		//
+		// 	// check that sharding was enabled in a previous state of the upstream cluster.
+		// 	if p.currentPSMDBSpec.Sharding.Enabled {
+		// 		// Config Server
+		// 		psmdb.Spec.Sharding.ConfigsvrReplSet.MultiAZ.Affinity = p.currentPSMDBSpec.Sharding.ConfigsvrReplSet.MultiAZ.Affinity
+		// 		// Proxy
+		// 		psmdb.Spec.Sharding.Mongos.MultiAZ.Affinity = p.currentPSMDBSpec.Sharding.Mongos.MultiAZ.Affinity
```

Comment:
Note: in theory, `p.currentPSMDBSpec.Sharding.ConfigsvrReplSet` and `p.currentPSMDBSpec.Sharding.Mongos` can be nil

Reply:
Yes, I know.
the important thing here is the following:
`applier.PodSchedulingPolicy()` shall be called after `applier.Engine()` and `applier.Proxy()` because exactly in those functions `spec..Sharding.ConfigsvrReplSet` and `Spec.Sharding.Mongos` are initialised